antidote-v1.0/MP1-syringe-redesign.md

```diff
@@ -0,0 +1,171 @@
+
+https://peter.bourgon.org/blog/2017/02/21/metrics-tracing-and-logging.html
+https://opentracing.io/specification/
+
+Not easy to debug stuff
+-It is really, really hard to see a user’s path through the system right now with flat logs. Debugging problems is nearly impossible right now, because of two main insufficiencies
+-Not enough context is given for each log message.
+-Flat logging won’t give us any insight whatsoever into how well the solution is scaling. This will be especially true if we fix Problem #1, as we’ll invest time into breaking Syringe apart into microservices but with flat logs we’ll be operating blind.
+Solution:
+-Logging is still good, but it needs to be structured, and it needs to be done in such a way that all the context is preserved, so we can filter on a UUID, for instance
+-Instrument the code for distributed tracing so that we can export traces to standardized distributed tracing tools like Jaeger and Honeycomb. This will allow us to take a piece of cardinal data (i.e. UUID) and see the end-to-end flow throughout the system.
+-OpenTracing is overwhelmingly the goto standard for distributed tracing, and is supported by numerous backends, including Jaeger, Honeycomb, Zipkin and more. https://opentracing.io/specification/
+
+Traces will begin on receipt of requests, either in GRPC (below) or rest-gateway if possible
+-https://medium.com/@masroor.hasan/tracing-infrastructure-with-jaeger-on-kubernetes-6800132a677
+-https://github.com/grpc-ecosystem/go-grpc-middleware/
+-NATS can also be instrumented: https://github.com/nats-io/not.go
+
+Spans
+-Api-to-scheduler
+-Scheduler-to-st2
+-Etc - need to build a trace diagram on the whiteboard and copy it here.
+-Links
+
+
+
+### Design Area - Lesson Networking
+
+The existing model works well but has shortcomings.
+In this project we should evaluate NetworkServiceMesh and figure out if it suits our needs.
+
+## Mini-Project Milestones
+
+This mini-project is too large to take place within a single platform release. To facilitate ongoing stability,
+and create logical steps to the end-goal, a set of milestones is listed below. These milestones should be done
+in the order shown, and can be added to any platform release that is deemed suitable.
+
+This document will be updated as these milestones are achieved, with links to the Pull Request(s) and/or
+that satisfy them.
+
+### MP1.1 - Move state to external database
+
+### MP1.2 - Break out garbage collection
+
+### MP1.3 - Break out TSDB
+
+### MP1.5 - Separate API and Scheduler and implement message queue
+
+### MP1.6 - Structured and centralized logging with Fluentd
+
+### MP1.7 - Observability instrumentation
+
+### MP1.8 - Syringe Security
+
+No need to go overboard here because we are running behind cloudflare in prod, but some common sense
+measures should be taken, especially since not everyone will be running behind cloudflare.
+
+Specifically things like rate limiting, source IP stuff, etc etc.
+
```

Comment:
AFAIU, here, you typically think of Denial of Service attacks.

I think there's also the concern of nasty beahviour by the users inside the labs, which I tried to describe in https://community.networkreliability.engineering/t/options-for-securing-linux-containers/140 ... maybe a different mini-project, as it relates to other tools, not exactly Syringe, but the fabrication of lab images ?

Reply:
This will be given it's own mini-project, as this is outside the scope of Syringe specifically. Stay tuned.

Reply:
I responded long ago in that forum topic, but just wanted to mention this is definitely part of MP3. So let's continue the conversation there.